replica/table.cc

```diff
@@ -3987,13 +4006,25 @@ future<> compaction_group::cleanup() {
         }
     };
 
+    auto permit = co_await _t.get_sstable_list_permit();
     auto updater = row_cache::external_updater(std::make_unique<compaction_group_cleaner>(*this));
 
     auto p_range = to_partition_range(token_range());
     tlogger.debug("Invalidating range {} for compaction group {} of table {} during cleanup.",
                   p_range, group_id(), _t.schema()->ks_name(), _t.schema()->cf_name());
+    // Since permit is still held, all actions below will be executed atomically:
     co_await _t._cache.invalidate(std::move(updater), p_range);
     _t._cache.refresh_snapshot();
+
+    co_await _t.delete_sstables_atomically(permit, _sstables_compacted_but_not_deleted);
```

Comment:
I'm worried that holding the permit around all deletions would be too long.
We could release the permit once the pending_delete log is written and the sstable set is updated in memory, and leave the actual deletion in the background, while holding the deletion gate.
That would be even more required once deletions are throttled.(#14940)

Also, should we use a rwlock rather than a simple semaphore so that snapshot and others can hold a read lock concurrently if they don't modify the sstables sets?